test/integration/apiserver/certreload/aggregator_certreload_test.go

```diff
@@ -0,0 +1,341 @@
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "api",
+		},
+		Subsets: []v1.EndpointSubset{
+			{
+				Addresses: []v1.EndpointAddress{{IP: aggregatorIP}},
+				Ports:     []v1.EndpointPort{{Port: int32(aggregatorPort), Protocol: v1.ProtocolTCP}},
+			},
+		},
+	}
+	_, err = kubeClient.CoreV1().Endpoints("kube-wardle").Create(context.Background(), ep, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+
+	_, err = aggregatorClient.ApiregistrationV1().APIServices().Create(context.TODO(), &apiregistrationv1.APIService{
+		ObjectMeta: metav1.ObjectMeta{Name: "v1alpha1.wardle.example.com"},
+		Spec: apiregistrationv1.APIServiceSpec{
+			Service: &apiregistrationv1.ServiceReference{
+				Namespace: "kube-wardle",
+				Name:      "api",
+			},
+			InsecureSkipTLSVerify: true, // this is okay as we don't want to validate wardle certificate
+			Group:                 "wardle.example.com",
+			Version:               "v1alpha1",
+			GroupPriorityMinimum:  200,
+			VersionPriority:       200,
+		},
+	}, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+
+	// wait until wardle is  available
+	if err = wait.Poll(300*time.Millisecond, 5*time.Second, func() (done bool, err error) {
+		svc, err := aggregatorClient.ApiregistrationV1().APIServices().Get(context.Background(), "v1alpha1.wardle.example.com", metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		for _, cond := range svc.Status.Conditions {
+			if cond.Type == apiregistrationv1.Available && cond.Status == apiregistrationv1.ConditionTrue {
+				return true, nil
+			}
+		}
+		return false, nil
+	}); err != nil {
+		return fmt.Errorf("failed waiting for wardle to become availiable, err = %v", err)
+	}
+
+	// check the are no discovery errors
+	err = wait.Poll(300*time.Millisecond, 5*time.Second, func() (done bool, err error) {
+		if _, err = kubeClient.Discovery().ServerResources(); err != nil {
+			t.Logf("discovery call returned an unexpected error %q, tyring one more time", err)
+			return false, nil
+		}
+
+		return true, nil
+	})
+	if err != nil {
+		return fmt.Errorf("failed waiting for discovery, err = %v", err)
+	}
+
+	// try to hit the new endpoint
+	res := kubeClient.RESTClient().Get().AbsPath("/apis/wardle.example.com/v1alpha1/flunders").Do(context.Background())
+	st := 0
+	res.StatusCode(&st)
+	if st != http.StatusOK {
+		return fmt.Errorf("unexpected status code %v while accessing the aggregated API server", st)
+	}
+
+	return nil
+}
+
+func swapProxyClientCertAndConnect(kubeClient *client.Clientset, proxyClientCertPath string, proxyClientKeyPath string, proxyClientKey []byte, proxyClientCert *x509.Certificate, expectedHTTPStatus int) error {
+	if err := ioutil.WriteFile(proxyClientCertPath, testutil.EncodeCertPEM(proxyClientCert), 0644); err != nil {
+		return err
+	}
+
+	if err := ioutil.WriteFile(proxyClientKeyPath, proxyClientKey, 0644); err != nil {
+		return err
+	}
+
+	// give it a time to reload the cert
+	time.Sleep(2 * time.Second)
```

Comment:
should you rather poll multiple times and wait for the expected result than just wait? From my experience, this usually leads to flaky tests. I know that the reload time is 1s so this SHOULD be enough but I would still rather retry and wait for couple of seconds than this.

Reply:
sure, good idea. I'll update the PR.